src/coreclr/vm/comdelegate.cpp

```diff
@@ -801,49 +893,66 @@ static Stub* SetupShuffleThunk(MethodTable* pDelMT, MethodDesc* pTargetMeth)
 
-    ShuffleThunkCache* pShuffleThunkCache = s_pShuffleThunkCache;
+    MethodDesc *pMD = pClass->GetInvokeMethod();
 
-    LoaderAllocator* pLoaderAllocator = pDelMT->GetLoaderAllocator();
-    if (pLoaderAllocator->IsCollectible())
+    // We haven't already setup a shuffle thunk, go do it now (which will cache the result automatically).
+    StackSArray<ShuffleEntry> rShuffleEntryArray;
+    if (GenerateShuffleArray(pMD, pTargetMeth, &rShuffleEntryArray))
     {
-        pShuffleThunkCache = ((AssemblyLoaderAllocator*)pLoaderAllocator)->GetShuffleThunkCache();
-    }
+        ShuffleThunkCache* pShuffleThunkCache = s_pShuffleThunkCache;
 
-    Stub* pShuffleThunk = pShuffleThunkCache->Canonicalize((const BYTE *)&rShuffleEntryArray[0]);
-    if (!pShuffleThunk)
-    {
-        COMPlusThrowOM();
-    }
+        LoaderAllocator* pLoaderAllocator = pDelMT->GetLoaderAllocator();
+        if (pLoaderAllocator->IsCollectible())
+        {
+            pShuffleThunkCache = ((AssemblyLoaderAllocator*)pLoaderAllocator)->GetShuffleThunkCache();
+        }
 
-    if (!pTargetMeth->IsStatic() && pTargetMeth->HasRetBuffArg() && IsRetBuffPassedAsFirstArg())
-    {
-        if (InterlockedCompareExchangeT(&pClass->m_pInstRetBuffCallStub, pShuffleThunk, NULL ) != NULL)
+        Stub* pShuffleThunk = pShuffleThunkCache->Canonicalize((const BYTE *)&rShuffleEntryArray[0]);
+        if (!pShuffleThunk)
         {
-            ExecutableWriterHolder<Stub> shuffleThunkWriterHolder(pShuffleThunk, sizeof(Stub));
-            shuffleThunkWriterHolder.GetRW()->DecRef();
-            pShuffleThunk = pClass->m_pInstRetBuffCallStub;
+            COMPlusThrowOM();
         }
+
+        if (isInstRetBuff)
+        {
+            if (InterlockedCompareExchangeT(&pClass->m_pInstRetBuffCallStub, pShuffleThunk, NULL ) != NULL)
+            {
+                ExecutableWriterHolder<Stub> shuffleThunkWriterHolder(pShuffleThunk, sizeof(Stub));
+                shuffleThunkWriterHolder.GetRW()->DecRef();
+                pShuffleThunk = pClass->m_pInstRetBuffCallStub;
+            }
+        }
+        else
+        {
+            if (InterlockedCompareExchangeT(&pClass->m_pStaticCallStub, pShuffleThunk, NULL ) != NULL)
+            {
+                ExecutableWriterHolder<Stub> shuffleThunkWriterHolder(pShuffleThunk, sizeof(Stub));
+                shuffleThunkWriterHolder.GetRW()->DecRef();
+                pShuffleThunk = pClass->m_pStaticCallStub;
+            }
+        }
+        return pShuffleThunk->GetEntryPoint();
     }
     else
     {
-        if (InterlockedCompareExchangeT(&pClass->m_pStaticCallStub, pShuffleThunk, NULL ) != NULL)
-        {
-            ExecutableWriterHolder<Stub> shuffleThunkWriterHolder(pShuffleThunk, sizeof(Stub));
-            shuffleThunkWriterHolder.GetRW()->DecRef();
-            pShuffleThunk = pClass->m_pStaticCallStub;
-        }
+#if defined(TARGET_RISCV64)
+        return CreateILDelegateShuffleThunk(pMD, pTargetMeth);
```

Comment:
Does the code for the IL-based shuffle thunks need to be cached similar to how the code for the non-IL based shuffle thunks is cached?

If I am reading the code correctly, we will create a new shuffle thunk for each delegate instance. It can manifest as a memory leak if the delegates needing these shuffle thunks are created in a loop.

Reply:
Yes, it creates thunk for each delegate instance now, thanks for forseeing the possible memory leak.

It can be cached on DelegateEEClass, the IL thunk would have to get the function address from the delegate object. Looks doable, I'll give it a try.